src/s3path.jl

```diff
@@ -0,0 +1,235 @@
+    g = FilePathsBase.Group()
+    blksize = 4096
+    blocks = 0
+    s = 0
+    last_modified = DateTime(0)
+
+    if exists(fp)
+        resp = s3_get_meta(fp.config, fp.bucket, fp.key)
+        # Example: "Thu, 03 Jan 2019 21:09:17 GMT"
+        last_modified = DateTime(
+            resp["Last-Modified"][1:end-4],
+            dateformat"e, d u Y H:M:S",
+        )
+        s = parse(Int, resp["Content-Length"])
+        blocks = ceil(Int, s / 4096)
+    end
+
+    return Status(0, 0, m, 0, u, g, 0, s, blksize, blocks, last_modified, last_modified)
+end
+
+# TODO: FilePathsBase should default to calling stat?
+FilePathsBase.lstat(fp::S3Path) = stat(fp)
+FilePathsBase.mode(fp::S3Path) = stat(fp).mode
+Base.size(fp::S3Path) = stat(fp).size
+FilePathsBase.created(fp::S3Path) = stat(fp).ctime
+FilePathsBase.modified(fp::S3Path) = stat(fp).mtime
+Base.islink(fp::S3Path) = false
+Base.issocket(fp::S3Path) = false
+Base.isfifo(fp::S3Path) = false
+Base.ischardev(fp::S3Path) = false
+Base.isblockdev(fp::S3Path) = false
+Base.ismount(fp::S3Path) = false
+# Need API for accessing object ACL permissions for this to work
+FilePathsBase.isexecutable(fp::S3Path) = false
+Base.isreadable(fp::S3Path) = true
+Base.iswritable(fp::S3Path) = true
+
+function Base.mkdir(fp::S3Path; recursive=false, exist_ok=false)
+    fp.dir || throw(ArgumentError("S3Path folders must end with '/': $fp"))
+
+    if exists(fp)
+        !exist_ok && error("$fp already exists.")
+    else
+        if hasparent(fp) && !exists(parent(fp))
+            if recursive
+                mkdir(parent(fp); recursive=recursive, exist_ok=exist_ok)
+            else
+                error(
+                    "The parent of $fp does not exist. " *
+                    "Pass recursive=true to create it."
+                )
+            end
+        end
+
+        write(fp, "")
+    end
+end
+
+function Base.rm(fp::S3Path; recursive=false, kwargs...)
+    if isdir(fp)
+        files = readpath(fp)
+
+        if recursive
+            for f in files
+                rm(f; recursive=recursive, kwargs...)
+            end
+        elseif length(files) > 0
+            error("S3 path $object is not empty. Use `recursive=true` to delete.")
+        end
+    end
+
+    @debug "delete: $fp"
+    s3_delete(fp.config, fp.bucket, fp.key)
+end
+
+function Base.readdir(fp::S3Path)
+    if isdir(fp)
+        k = fp.key
+        # Only list the files and "dirs" within this S3 "dir"
+        objects = s3_list_objects(fp.config, fp.bucket, k; delimiter="")
+
+        # Only list the basename and not the full key
+        basenames = Set(s3_get_name(k, string(o["Key"])) for o in objects)
+
+        # Lexographically sort the results
+        return sort!(filter!(!isempty, collect(basenames)))
+    else
+        throw(ArgumentError("\"$fp\" is not a directory"))
+    end
+end
+
+Base.read(fp::S3Path) = s3_get(fp.config, fp.bucket, fp.key)
+Base.read(fp::S3Path, ::Type{String}) = String(read(fp))
```

Comment:
We can probably get rid of this?
Since the new syntax is to hust call `String(read)`?

Or if we must keep this then we should just define this in `FilePathsBase`

Reply:
I didn't realize `read(path, String)` was being deprecated.

Reply:
I think it is gone

Reply:
👌